### PR TITLE
GH-38157: [C++][Parquet] Decode: PlainBooleanDecoder batch decode

### LINKED ISSUE #38157: [C++][Parquet] Parquet: PlainBooleanDecoder batch decode
### Describe the enhancement requested

Using Batch Decode to optimize `PlainBooleanDecoder`.

### Component(s)

C++, Parquet

### PR DESCRIPTION
**Pull request by mapleFU**
### Rationale for this change

Using batch-decode to optimize PlainBooleanDecoder.

### What changes are included in this PR?

Optimize PlainBooleanDecoder.

### Are these changes tested?

Already has

### Are there any user-facing changes?

no

* Closes: #38157

### CODE REVIEW

File: cpp/src/parquet/encoding.cc [hunk 1]
```diff
@@ -1156,16 +1160,21 @@ inline int PlainBooleanDecoder::DecodeArrow(
 
 int PlainBooleanDecoder::Decode(uint8_t* buffer, int max_values) {
   max_values = std::min(max_values, num_values_);
-  bool val;
+  constexpr int kBatchSize = 1024;
+  std::array<bool, kBatchSize> bit_read_scratch;
   ::arrow::internal::BitmapWriter bit_writer(buffer, 0, max_values);
```
Comment (mapleFU) [inline cpp/src/parquet/encoding.cc:1165]: In fact I don't know the differences between bitWriter and `FirstTimeBitWriter`... Would the later one be better? 🤔

**Comment by mapleFU** [comment]
@pitrou @wgtmac

File: cpp/src/parquet/encoding.cc (hunk 1 above)
Comment (mapleFU) [reply cpp/src/parquet/encoding.cc:1165]: And in fact the fastest the way is directly copy LSB bits to BitWriter 🤔

File: cpp/src/parquet/encoding.cc [hunk 2]
```diff
@@ -1156,16 +1160,21 @@ inline int PlainBooleanDecoder::DecodeArrow(
...
+    int batch_size = std::min(max_values - i, kBatchSize);
+    if (bit_reader_->GetBatch(/*num_bits=*/1, bit_read_scratch.data(), batch_size) !=
+        batch_size) {
       ParquetException::EofException();
     }
-    if (val) {
-      bit_writer.Set();
+    for (int j = 0; j < batch_size; ++j) {
+      if (bit_read_scratch[j]) {
+        bit_writer.Set();
```
Comment (wgtmac) [inline cpp/src/parquet/encoding.cc:1174]: It seems that bit writer has the chance to accept batch write.

Comment (mapleFU) [reply cpp/src/parquet/encoding.cc:1174]: Just like mentioned here: https://github.com/apache/arrow/pull/38158#discussion_r1350521129

Maybe directly convert the LSB input to output is fastest.

**Comment by mapleFU** [comment]
@pitrou do you have some advice for this? Since maybe introduce BitWriter and write directly is also ok, but I didn't know if there're some usable tools here...

**Comment by pitrou** [comment]
Do you actually use `PlainBooleanDecoder::Decode` or is `DecodeArrow` more useful?

**Comment by mapleFU** [comment]
DecodeArrow is useless, since `RecordReader` always prefer calling `Decode` and `DecodeSpace`

**Comment by pitrou** [comment]
Does it call `Decode(uint8_t*)` or `Decode(bool*)`?

**Comment by mapleFU** [comment]
@pitrou uint8_t sadly :-(

**Comment by pitrou** [comment]
In any case, I'm not sure `BitReader` is useful at all here. You could use `CopyBitmap` directly.

**Comment by mapleFU** [comment]
I've gothrough the code, the main reason that it's slow is that we don't have a fast enough `unpack8`...

**Pull request closed without merging**